File: streamflowSignatures/julia/src/pulses.jl
"""
Pulse metrics signatures.

High/low pulse counts, durations, TQmean, and flow reversals.
"""

# Season definitions
const SEASONS = Dict(
    "winter" => [12, 1, 2],
    "spring" => [3, 4, 5],
    "summer" => [6, 7, 8],
    "fall" => [9, 10, 11]
)


"""
    count_pulses(Q::Vector, threshold::Real, above::Bool) -> (count, mean_duration)

Count pulse events and calculate mean duration.

Parameters
----------
Q : Vector
    Daily discharge values
threshold : Real
    Threshold value for pulse detection
above : Bool
    If true, count pulses above threshold; if false, below

Returns
-------
Tuple{Int, Float64}
    (number of pulses, mean duration in days)
"""
function count_pulses(Q::AbstractVector{<:Real}, threshold::Real, above::Bool)
    n = length(Q)
    if n == 0
        return (0, NaN)
    end

    # Determine which days are in pulse state
    if above
        in_pulse = Q .> threshold
    else
        in_pulse = Q .< threshold
    end

    # Handle NaN - not in pulse
    in_pulse[isnan.(Q)] .= false

    # Count pulse events and durations
    pulse_count = 0
    durations = Int[]
    current_duration = 0

    for i in 1:n
        if in_pulse[i]
            current_duration += 1
        else
            if current_duration > 0
                pulse_count += 1
                push!(durations, current_duration)
                current_duration = 0
            end
        end
    end

    # Handle pulse at end of series
    if current_duration > 0
        pulse_count += 1
        push!(durations, current_duration)
    end

    mean_duration = isempty(durations) ? NaN : mean(durations)

    return (pulse_count, mean_duration)
end


"""
    count_flow_reversals(Q::Vector; threshold_pct=0.02) -> Int

Count direction changes in flow that exceed a magnitude threshold.

A reversal occurs when flow changes from increasing to decreasing or vice versa,
and the magnitude of the change exceeds threshold_pct of the current flow value.

Parameters
----------
Q : Vector
    Daily discharge values
threshold_pct : Float64
    Minimum magnitude threshold as fraction of current flow (default 0.02 = 2%)
"""
function count_flow_reversals(Q::AbstractVector{<:Real}; threshold_pct::Float64=0.02)
    # Preprocessor guarantees no NaNs in valid years
    n = length(Q)
    if n < 3
        return 0
    end

    Q_interp = Float64.(Q)

    reversals = 0

    # Match R/Python implementation: check prev_change and next_change
    # Only count reversal if |next_change| > threshold
    @inbounds for i in 2:(n-1)
        prev_change = Q_interp[i] - Q_interp[i-1]
        next_change = Q_interp[i+1] - Q_interp[i]
        threshold = abs(threshold_pct * Q_interp[i])

        if abs(next_change) > threshold
            # Peak (rising to falling)
            if prev_change >= 0 && next_change < 0
                reversals += 1
            # Trough (falling to rising)
            elseif prev_change <= 0 && next_change > 0
                reversals += 1
            end
        end
    end

    return reversals
end


"""
    calculate_tqmean(Q::Vector) -> Float64

Calculate percentage of days with flow above mean.
"""
function calculate_tqmean(Q::AbstractVector{<:Real})
    Q_valid = filter(!isnan, Q)
    if isempty(Q_valid)
        return NaN
    end

    mean_Q = mean(Q_valid)
    days_above = sum(Q_valid .> mean_Q)

    return days_above / length(Q_valid) * 100
end


"""
    calculate_pulse_metrics(df::DataFrame) -> Dict

Calculate pulse metric signatures with trend statistics.

Calculates 14 metrics:
- n_high_pulses_year, n_low_pulses_year: Pulse counts (year-specific thresholds)
- dur_high_pulses_year, dur_low_pulses_year: Mean pulse durations (year-specific)
- n_high_pulses_all, n_low_pulses_all: Pulse counts (period-of-record thresholds)
- dur_high_pulses_all, dur_low_pulses_all: Mean pulse durations (period-of-record)
- TQmean: Percentage of days above mean
- Flow_Reversals_annual: Total annual reversals
- Flow_Reversals_winter/spring/summer/fall: Seasonal reversals

The *_year metrics use year-specific Q90/Q10 thresholds, while *_all metrics
use Q90/Q10 calculated over the entire period of record.

Each metric produces 8 statistics via generate_stats().

Parameters
----------
df : DataFrame
    Daily streamflow data with columns: water_year, Q, month, dowy
    (preprocessor guarantees all years are valid)

Returns
-------
Dict{String, Float64}
    Dictionary of signature statistics
"""
function calculate_pulse_metrics(df::DataFrame; trend_completeness::Union{Nothing, Float64}=nothing, decade_completeness::Union{Nothing, Float64}=nothing, min_values_for_stats::Union{Nothing, Int}=nothing, changepoint::Union{Nothing, NamedTuple}=nothing, collector::Union{Nothing, AnnualCollector}=nothing)
    result = Dict{String, Float64}()

    metrics = [
        "n_high_pulses_year", "n_low_pulses_year",
        "dur_high_pulses_year", "dur_low_pulses_year",
        # Period-of-record threshold metrics (matching R's *_all metrics)
        "n_high_pulses_all", "n_low_pulses_all",
        "dur_high_pulses_all", "dur_low_pulses_all",
        "TQmean", "Flow_Reversals_annual",
        "Flow_Reversals_winter", "Flow_Reversals_spring",
        "Flow_Reversals_summer", "Flow_Reversals_fall"
    ]

    # Column validation
    valid, missing_cols = validate_columns(df, ["Q", "water_year", "month", "dowy"])
    if !valid
        @warn "calculate_pulse_metrics: Missing columns: $missing_cols"
        for m in metrics
            merge!(result, empty_stats(m))
        end
        return result
    end

    if nrow(df) == 0
        for m in metrics
            merge!(result, empty_stats(m))
        end
        return result
    end

    # Calculate period-of-record thresholds for *_all metrics
    Q_all = coalesce_q(df.Q)
    Q_all_valid = filter(!isnan, Q_all)
    if length(Q_all_valid) < 30
        for m in metrics
            merge!(result, empty_stats(m))
        end
        return result
    end
    q90_all = quantile(Q_all_valid, 0.90)
    q10_all = quantile(Q_all_valid, 0.10)

    years = unique(df.water_year)
    annual_data = DataFrame(
        water_year = years,
        n_high_pulses_year = fill(NaN, length(years)),
        n_low_pulses_year = fill(NaN, length(years)),
        dur_high_pulses_year = fill(NaN, length(years)),
        dur_low_pulses_year = fill(NaN, length(years)),
        n_high_pulses_all = fill(NaN, length(years)),
        n_low_pulses_all = fill(NaN, length(years)),
        dur_high_pulses_all = fill(NaN, length(years)),
        dur_low_pulses_all = fill(NaN, length(years)),
        TQmean = fill(NaN, length(years)),
        Flow_Reversals_annual = fill(NaN, length(years)),
        Flow_Reversals_winter = fill(NaN, length(years)),
        Flow_Reversals_spring = fill(NaN, length(years)),
        Flow_Reversals_summer = fill(NaN, length(years)),
        Flow_Reversals_fall = fill(NaN, length(years))
    )

    for yr in years
        year_mask = df.water_year .== yr
        if !any(skipmissing(year_mask))
            continue
        end
        year_df = df[coalesce.(year_mask, false), :]
        year_df = sort(year_df, :dowy)

        # Convert Q to clean Float64 vector
        Q_year = coalesce_q(year_df.Q)

        # FIX: Calculate year-specific thresholds (matching R/Python)
        Q_valid = filter(!isnan, Q_year)
        if length(Q_valid) < 30
            continue
        end
        high_threshold = quantile(Q_valid, 0.90)
        low_threshold = quantile(Q_valid, 0.10)

        yr_idx = findfirst(annual_data.water_year .== yr)
        yr_idx === nothing && continue

        # High and low pulses using year-specific thresholds
        n_high, dur_high = count_pulses(Q_year, high_threshold, true)
        n_low, dur_low = count_pulses(Q_year, low_threshold, false)

        annual_data[yr_idx, :n_high_pulses_year] = n_high
        annual_data[yr_idx, :n_low_pulses_year] = n_low
        annual_data[yr_idx, :dur_high_pulses_year] = dur_high
        annual_data[yr_idx, :dur_low_pulses_year] = dur_low

        # High and low pulses using period-of-record thresholds (*_all metrics)
        n_high_all, dur_high_all = count_pulses(Q_year, q90_all, true)
        n_low_all, dur_low_all = count_pulses(Q_year, q10_all, false)

        annual_data[yr_idx, :n_high_pulses_all] = n_high_all
        annual_data[yr_idx, :n_low_pulses_all] = n_low_all
        annual_data[yr_idx, :dur_high_pulses_all] = dur_high_all
        annual_data[yr_idx, :dur_low_pulses_all] = dur_low_all

        # TQmean
        annual_data[yr_idx, :TQmean] = calculate_tqmean(Q_year)

        # Flow reversals - annual
        annual_data[yr_idx, :Flow_Reversals_annual] = count_flow_reversals(Q_year)

        # Seasonal reversals
        for (season, months) in SEASONS
            month_mask = [coalesce(m, 0) in months for m in year_df.month]
            if sum(month_mask) >= 30
                season_Q = Q_year[month_mask]
                annual_data[yr_idx, Symbol("Flow_Reversals_$season")] = count_flow_reversals(season_Q)
            end
        end
    end

    if nrow(annual_data) < 3
        for m in metrics
            merge!(result, empty_stats(m))
        end
        return result
    end

    # Generate statistics for all metrics
    result = generate_stats(annual_data; value_cols=metrics, trend_completeness=trend_completeness, decade_completeness=decade_completeness, min_values_for_stats=min_values_for_stats, changepoint=changepoint, collector=collector)

    return result
end


"""
    calculate_negative_days(df; trend_completeness=nothing, decade_completeness=nothing)

Count days with negative streamflow (Q < 0) per water year.
"""
function calculate_negative_days(df::DataFrame; trend_completeness::Union{Nothing, Float64}=nothing, decade_completeness::Union{Nothing, Float64}=nothing, min_values_for_stats::Union{Nothing, Int}=nothing, changepoint::Union{Nothing, NamedTuple}=nothing, collector::Union{Nothing, AnnualCollector}=nothing)
    result = Dict{String, Float64}()

    valid, missing_cols = validate_columns(df, ["Q", "water_year"])
    if !valid
        @warn "calculate_negative_days: Missing columns: $missing_cols"
        merge!(result, empty_stats("negative_ann"))
        return result
    end

    annual_data = combine(groupby(df, :water_year),
        :Q => (q -> sum(x -> !isnan(x) && x < 0, q)) => :negative_ann
    )

    result = generate_stats(annual_data; value_cols=["negative_ann"],
        trend_completeness=trend_completeness, decade_completeness=decade_completeness, min_values_for_stats=min_values_for_stats, changepoint=changepoint, collector=collector)

    return result
end

File: streamflowSignatures/rpkg/R/pulses.R
#' Identify pulse events (runs above/below threshold)
#' @keywords internal
identify_pulses <- function(flow_vector, threshold, above = TRUE) {
  exceeds <- if (above) flow_vector > threshold else flow_vector < threshold
  exceeds[is.na(exceeds)] <- FALSE

  runs <- rle(exceeds)
  pulse_lengths <- runs$lengths[runs$values == TRUE]

  if (length(pulse_lengths) > 0) {
    list(n_pulses = length(pulse_lengths), mean_duration = mean(pulse_lengths))
  } else {
    list(n_pulses = 0L, mean_duration = NA_real_)
  }
}

#' Count flow reversals
#' @keywords internal
count_flow_reversals <- function(flow_vector, threshold_pct = NULL) {
  if (is.null(threshold_pct)) threshold_pct <- pkg_env$flow_reversal_threshold

  stopifnot(!any(is.na(flow_vector)))
  # No NA-cleaning happens here (the stopifnot above guarantees none);
  # the "_clean" name is historical.
  flow_clean <- flow_vector

  n <- length(flow_clean)
  if (n < 3) return(0L)

  reversal_count <- 0L
  for (i in 2:(n - 1)) {
    prev_change <- flow_clean[i] - flow_clean[i - 1]
    next_change <- flow_clean[i + 1] - flow_clean[i]
    threshold   <- abs(threshold_pct * flow_clean[i])

    if (abs(next_change) > threshold) {
      if (prev_change >= 0 && next_change < 0) reversal_count <- reversal_count + 1L
      else if (prev_change <= 0 && next_change > 0) reversal_count <- reversal_count + 1L
    }
  }
  reversal_count
}

#' Calculate pulse metrics
#'
#' Computes high/low pulse counts and durations (using year-specific and
#' period-of-record thresholds), TQmean, and annual/seasonal flow reversals.
#'
#' @param streamflow_data A data.frame with columns: water_year, Q, dowy, month.
#' @return Named list of 8 statistics per pulse metric (14 metrics).
#' @export
calculate_pulse_metrics <- function(streamflow_data,
                                   trend_completeness = NULL,
                                   decade_completeness = NULL, min_values_for_stats = NULL, changepoint = NULL, collector = NULL) {
  required_cols <- c("water_year", "Q", "dowy", "month")
  missing <- setdiff(required_cols, colnames(streamflow_data))
  if (length(missing) > 0) stop(paste("Missing required columns:", paste(missing, collapse = ", ")))

  q90_all <- quantile(streamflow_data$Q, probs = 0.90, na.rm = TRUE)
  q10_all <- quantile(streamflow_data$Q, probs = 0.10, na.rm = TRUE)

  years <- unique(streamflow_data$water_year)

  metric_names <- c("n_high_pulses_year", "n_low_pulses_year",
                    "n_high_pulses_all", "n_low_pulses_all",
                    "dur_high_pulses_year", "dur_low_pulses_year",
                    "dur_high_pulses_all", "dur_low_pulses_all",
                    "TQmean",
                    "Flow_Reversals_annual", "Flow_Reversals_winter",
                    "Flow_Reversals_spring", "Flow_Reversals_summer",
                    "Flow_Reversals_fall")

  pulse_metrics <- data.frame(water_year = years)
  for (nm in metric_names) pulse_metrics[[nm]] <- NA_real_

  for (yr in years) {
    year_data <- streamflow_data[streamflow_data$water_year == yr, ]
    year_data <- year_data[order(year_data$dowy), ]

    q90_year <- quantile(year_data$Q, probs = 0.90, na.rm = TRUE)
    q10_year <- quantile(year_data$Q, probs = 0.10, na.rm = TRUE)
    if (is.na(q90_year) || is.na(q10_year)) next

    hp_yr  <- identify_pulses(year_data$Q, q90_year, above = TRUE)
    lp_yr  <- identify_pulses(year_data$Q, q10_year, above = FALSE)
    hp_all <- identify_pulses(year_data$Q, q90_all,  above = TRUE)
    lp_all <- identify_pulses(year_data$Q, q10_all,  above = FALSE)

    annual_mean <- mean(year_data$Q, na.rm = TRUE)
    days_above  <- sum(year_data$Q > annual_mean, na.rm = TRUE)
    total_valid <- sum(!is.na(year_data$Q))
    tqmean      <- (days_above / total_valid) * 100

    annual_rev <- count_flow_reversals(year_data$Q)

    season_months <- list(winter = c(12, 1, 2), spring = c(3, 4, 5),
                          summer = c(6, 7, 8), fall = c(9, 10, 11))
    seasonal_rev <- vapply(season_months, function(mos) {
      sub <- year_data[year_data$month %in% mos, ]
      if (nrow(sub) >= 30) count_flow_reversals(sub$Q) else NA_real_
    }, numeric(1))

    idx <- which(pulse_metrics$water_year == yr)
    pulse_metrics$n_high_pulses_year[idx]    <- hp_yr$n_pulses
    pulse_metrics$n_low_pulses_year[idx]     <- lp_yr$n_pulses
    pulse_metrics$n_high_pulses_all[idx]     <- hp_all$n_pulses
    pulse_metrics$n_low_pulses_all[idx]      <- lp_all$n_pulses
    pulse_metrics$dur_high_pulses_year[idx]  <- hp_yr$mean_duration
    pulse_metrics$dur_low_pulses_year[idx]   <- lp_yr$mean_duration
    pulse_metrics$dur_high_pulses_all[idx]   <- hp_all$mean_duration
    pulse_metrics$dur_low_pulses_all[idx]    <- lp_all$mean_duration
    pulse_metrics$TQmean[idx]                <- tqmean
    pulse_metrics$Flow_Reversals_annual[idx] <- annual_rev
    pulse_metrics$Flow_Reversals_winter[idx] <- seasonal_rev["winter"]
    pulse_metrics$Flow_Reversals_spring[idx] <- seasonal_rev["spring"]
    pulse_metrics$Flow_Reversals_summer[idx] <- seasonal_rev["summer"]
    pulse_metrics$Flow_Reversals_fall[idx]   <- seasonal_rev["fall"]
  }

  generate_stats(pulse_metrics, value_cols = metric_names, year_col = "water_year",
                 trend_completeness = trend_completeness,
                 decade_completeness = decade_completeness, min_values_for_stats = min_values_for_stats, changepoint = changepoint, collector = collector)
}

#' Count days with negative streamflow per water year
#'
#' Counts the number of days with Q < 0 in each water year, then produces
#' 8 trend statistics via \code{generate_stats()}.
#'
#' @param streamflow_data A data.frame with columns: water_year, Q.
#' @param trend_completeness Forwarded to generate_stats.
#' @param decade_completeness Forwarded to generate_stats.
#' @return Named list with 8 statistics for negative_ann.
#' @export
calculate_negative_days <- function(streamflow_data,
                                    trend_completeness = NULL,
                                    decade_completeness = NULL, min_values_for_stats = NULL, changepoint = NULL, collector = NULL) {
  required_cols <- c("water_year", "Q")
  missing <- setdiff(required_cols, colnames(streamflow_data))
  if (length(missing) > 0) {
    warning(paste("calculate_negative_days: Missing columns:", paste(missing, collapse = ", ")))
    return(empty_stats("negative_ann"))
  }

  annual <- aggregate(Q ~ water_year, data = streamflow_data,
                      FUN = function(q) sum(!is.na(q) & q < 0))
  names(annual)[2] <- "negative_ann"

  generate_stats(annual, value_cols = "negative_ann", year_col = "water_year",
                 trend_completeness = trend_completeness,
                 decade_completeness = decade_completeness, min_values_for_stats = min_values_for_stats, changepoint = changepoint, collector = collector)
}
